### ISSUE TITLE
Azure Blob using service principal and oauth2

### ISSUE THREAD

**Issue by fdefilippo** [rclone/rclone#3230]
#### What is your current rclone version (output from `rclone version`)?

rpm -q rclone 
rclone-1.47.0-1

#### What problem are you are trying to solve?

accessing azure blob using oauth2 and service principal

#### How do you think rclone should be changed to solve that?

Access azureblob using oauth2 to get an access token owning client_id and client_key:

```
POST https://login.microsoftonline.com/{{tenant_id}}/oauth2/token

body:
client_id: xxxxx (id_service principal)
client_secret: XXXXXXXXXXXXXXXXXXXXXXX
resource: https://storage.azure.com/
grant_type: client_credentials
```

below a simple shell script client:

```sh
#!/bin/bash

client_id="xxx"
client_secret="yyy"
app_id="zzz"
storage_account="myaccount"
container_name="backup"

# url encoded resource
resource="https%3A%2F%2Fstorage.azure.com%2F"

curl_opts='-H "Content-Type: application/x-www-form-urlencoded"'
mthd="https"

# Access Token
json=$(curl -s -X POST -d "grant_type=client_credentials&resource=${resource}&client_id=${client_id}&client_secret=${client_secret}" "${mthd}://login.microsoftonline.com/${app_id}/oauth2/token?api-version=1.0")

if [[ $? -gt 0 || ${#json} -lt 80 ]]; then
echo "error getting token"
exit
fi

access_token=$(awk '/access_token/ {if (match($0,/"access_token":"([^"]+)/,at)) print at[1];}' <<< $json)
token_type=$(awk '/token_type/ {if (match($0,/"token_type":"([^"]+)/,tt)) print tt[1];}' <<< $json)

body=$(curl -s \
-H "Authorization: $token_type $access_token" \
-H 'x-ms-blob-type: BlockBlob' \
-H 'x-ms-version: 2018-03-28' \
"${mthd}://${storage_account}.blob.core.windows.net/${container_name}?restype=container&comp=list")
```

**Comment by ncw** [comment]
I think these are the relevant MS docs: https://docs.microsoft.com/en-us/azure/storage/common/storage-auth-aad-app

rclone has an oauth framework which we could use to implement this I think.  This will handle renewing the token too.

Do you want to work on this @fdefilippo 

@sandeepkru any other thoughts?

**Comment by fdefilippo** [comment]
> I think these are the relevant MS docs: https://docs.microsoft.com/en-us/azure/storage/common/storage-auth-aad-app
> 
> rclone has an oauth framework which we could use to implement this I think. This will handle renewing the token too.
> 
> Do you want to work on this @fdefilippo

I would really like it unfortunately I don't have the necessary skills. thank you.

**Comment by sandeepkru** [comment]
The approach should work, will dig more details, may be we can solve other open issue also if it works.

**Comment by musiczhzhao** [comment]
Hello, is there any recent update on this feature?

**Comment by ncw** [comment]
Do you want to work on it @musiczhzhao ?

**Comment by musiczhzhao** [comment]
> Do you want to work on it @musiczhzhao ?

Hi I wish I can but I don't think I have the experience/expertise for this task. I thought @sandeepkru was looking into this?

**Comment by viyh** [comment]
+1, this is necessary these days.

**Comment by ncw** [comment]
@viyh I'm looking for a volunteer for this... The GCS backend does this already allowing tokens or oauth and rclone has a framework for oauth so it shouldn't be too tricky if you wanted to have a go...

**Comment by viyh** [comment]
@ncw Unfortunately, I'm not well enough versed in Go to be able to do this. :(

**Comment by jimjh** [comment]
@ncw May I get some feedback on https://github.com/rclone/rclone/pull/4848? I am still working unit tests, but would like your opinion on the overall design.

**Comment by ncw** [comment]
Thanks to @jmjh support for this has been merged to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54

To use, install the beta and look at the advanced config for the azureblob backend.

**Comment by chscott** [comment]
@ncw @jmjh Can this be added to OneDrive/SharePoint as well? The same need to obtain access tokens via Graph's client_credentials flow exists there, too. Graph does not provide a refresh token when using the client_credentials flow.
